### PR TITLE
Josh Adelie Changes

### PR DESCRIPTION
**Pull request by jrich129**
Bug Fixes
-Fixed standardize on matrix.py to work with NumPy arrays
-Corrected the sklearn score function for R-Squared
-Fixed a typo in the centers and scales for the standardize class
-Corrected a minor typo in the y-axis label of the deviance profile plot
-Updated link in glm.ipynb to the correct API reference for GlmBase64

New Additions
-Changed cv.py to show average deviance rather than average loss
-Updated cv.py to include two vertical dashed lines in the _plot_loss function 
-Updated fit function in cv.py to reuse the same lambda path as the original fit when refitting
-Edited Cox to have start default to a vector of zero
-Added plot_validation_curve method to sklearn GroupElasticNet for visualizing R^2 and active set size
-Added center and scales to standardize docstring 
-Added nonzero coefficients to diagnostic plots
-Made deviance a value included in the CVgprnet output

### CODE REVIEW

File: adelie/cv.py [hunk 1]
```diff
@@ -43,16 +43,55 @@ class CVGrpnetResult:
...
+        """Plots the average K-fold CV deviance.
 
-        For each fitted :math:`\\lambda`, the average K-fold CV loss
+        For each fitted :math:`\\lambda`, the average K-fold CV deviance
         as well as an error bar of one standard deviation (above and below) is plotted.
+        
+        Parameters
+        ----------
+        glm : Union[GlmBase32, GlmBase64, GlmMultiBase32, GlmMultiBase64], optional
+            GLM object needed to compute deviance. If None, uses the stored GLM object.
```
Comment (JamesYang007) [inline adelie/cv.py]: For consistency, Python expressions should be placed within double backticks like ``None`` (same comment for the next 3 lines):
```
If ``None``, uses the stored GLM object.
Default is ``None``.
```
Also, add `None` in the `Union`

File: adelie/diagnostic.py [hunk 2]
```diff
@@ -1329,16 +1377,21 @@ def plot_coefficients(self, **kwargs):
...
-    def plot_devs(self):
+    def plot_devs(self, **kwargs):
         """Plots the deviance profile.
 
         See Also
         --------
         adelie.diagnostic.plot_devs
         """
+        # Compute number of nonzero coefficients at each lambda
+        dof = np.sum(self.betas != 0, axis=1)
```
Comment (JamesYang007) [inline adelie/diagnostic.py]: `nnz_counts` doesn't seem to be used?

File: adelie/diagnostic.py [hunk 3]
```diff
@@ -1184,16 +1227,21 @@ def plot_coefficients(self, **kwargs):
...
+        nnz_counts = np.asarray(dof).ravel()
```
Comment (JamesYang007) [inline adelie/diagnostic.py]: Same here - doesn't seem to be used

File: adelie/diagnostic.py [hunk 4]
```diff
@@ -711,12 +719,28 @@ def plot_coefficients(
...
+        ax2.set_xlabel("Number of non-zero coefficients")
+
     return fig, ax
 
 
 def plot_devs(
     lmdas: np.ndarray,
     devs: np.ndarray,
+    *,
+    betas: csr_matrix=None,
```
Comment (JamesYang007) [inline adelie/diagnostic.py]: ``Union[csr_matrix, None] =None`` (similar change in the docstring)

File: adelie/cv.py [hunk 5]
```diff
@@ -111,13 +169,24 @@ def fit(
         )
         logger.logger.setLevel(logger_level)
 
-        lmda_path_size = 100
+        # lmda_path_size = 100
```
Comment (JamesYang007) [inline adelie/cv.py]: Shall we just delete this?

File: adelie/sklearn.py [hunk 6]
```diff
@@ -231,10 +233,10 @@ def score(self, X: np.ndarray, y: np.ndarray) -> float:
             The R-squared score.
         """
         yhat = self.predict(X)
-        ybar = np.mean(y)
+        ybar = np.mean(self.glm_.y * self.glm_.weights)
```
Comment (JamesYang007) [inline adelie/sklearn.py]: should be `np.sum` since weights are already normalized to sum to 1.

File: adelie/sklearn.py [hunk 7]
```diff
@@ -249,6 +251,84 @@ def _validate_params(self):
...
+        # If predictions is 2D (n_lambdas, n_samples), axis=1 is correct
+        # If predictions is 1D, wrap in a 2D array for apply_along_axis
+        if predictions.ndim == 1:
+            predictions = predictions[np.newaxis, :]
+        r2vec = np.apply_along_axis(lambda yhat: r2_score(y_val, yhat), axis=1, arr=predictions)
+
+        lam_path = self.lambda_
+        beta_0 = self.intercept_
+        # Handle sparse or dense coef_
+        betas = self.coef_.toarray() if hasattr(self.coef_, "toarray") else np.asarray(self.coef_)
```
Comment (JamesYang007) [inline adelie/sklearn.py]: I would prefer to perform this check by doing `isinstance`. It's more explicit about what we expect. It'll either be `csr_matrix` or `np.ndarray` right?

**Review by JamesYang007 — CHANGES_REQUESTED** [review]
Minor comments. Excellent work!

File: adelie/cv.py (hunk 5 above)
Comment (jrich129) [reply adelie/cv.py]: Yeah I think it's fine to delete it

File: adelie/diagnostic.py (hunk 2 above)
Comment (jrich129) [reply adelie/diagnostic.py]: Yes, you are right - I removed them. 

In the plot_devs function (like line 738) we say that the user can specify if they want to have the top axis show number of non-zero coeffs. Right now there is no functionality to turn that off, so I fixed that by changing the parameter 'betas' they can pass in.

File: adelie/diagnostic.py (hunk 3 above)
Comment (jrich129) [reply adelie/diagnostic.py]: Yup!

**Comment by jwijffels** [comment]
+1 to have this integrated, that would make the plotting facilities of cv grpnet with the deviance similar to the r package
